### PR TITLE
Add DataFrame and Arrow export for mesh data 

### PR DESCRIPTION
**Pull request by banesullivan**
## Summary

Add `to_pandas()`, `to_arrow()`, and `__arrow_c_stream__()` on `DataSet`, `DataSetAttributes`, and `Table`. This lets users move mesh arrays easily into pandas, polars, DuckDB, or pyarrow. Or browse mesh data in your IDEs data explorer (e.g., Positron's Data Explorer)

Companion to #8582. Addresses the output side of #5516 and #265.

## Motivation

If you want a pandas data frame today, you have to hand roll this everytime"

```python
df = pd.DataFrame({k: v for k, v in mesh.point_data.items() if v.ndim == 1})
```

That drops multi-component fields (velocity, RGBA, tensors), fails on `field_data`, and has no path to Arrow-native consumers like polars or DuckDB.

The UX here just hasn't been great

## Benefits

- IDE variable inspection. `mesh.point_data.to_pandas()` opens directly in Positron's Data Explorer, VS Code's Jupyter data viewer, and JupyterLab. Sort columns, filter rows, inspect dtypes, view summary statistics.
- interop with more of the Python data sci stack. `pa.table(mesh)`, `pl.from_arrow(pa.table(mesh))`, `duckdb.sql("FROM mesh WHERE pressure > 100")`. Zero copies across the boundary thanks to the Arrow PyCapsule interface.
- One-liner export (of data, not the mesh!): `mesh.to_pandas().to_parquet(...)`, `.to_csv(...)`, `.to_sql(...)`. Every pandas IO backend is available for handoff to dashboards or downstream analytics.
- Ergonomics for DataFrame-first users. pandas and polars users can reach for familiar APIs like `df.describe()`, `df.nlargest(10, 'stress')`, or `df.groupby(...)` instead of learning a second filtering idiom on the pyvista dataset.
- Tutorials: "Here is your data as a table" is more approachable than "here is a dict-like array wrapper" for newcomers to scientific computing.

## Details

- `DataSetAttributes.to_pandas()` and `.to_arrow()` export point_data or cell_data row-aligned. Multi-component arrays expand to one column per component, named `{array_name}_{i}`. Tensors flatten the trailing dims in C order.
- `DataSetAttributes.__arrow_c_stream__(requested_schema=None)` implements the Arrow PyCapsule interface so pyarrow, polars, DuckDB, ibis, and narwhals ingest with zero copies.
- `DataSet.to_pandas(association)` and `.to_arrow(association)` delegate to `point_data` or `cell_data`. The `association` parameter accepts the string form (`'point'`, `'cell'`) or the `FieldAssociation` enum.
- `DataSet.__arrow_c_stream__` defaults to `point_data`, so `pa.table(mesh)` works out of the box.
- `Table.to_arrow()` and `Table.__arrow_c_stream__()` mirror the same interface.
- Memory sharing is path-dependent. `to_arrow` / `__arrow_c_stream__` wrap 1D contiguous numeric columns zero-copy; expanded multi-component columns (`{name}_{i}`) are strided slices and get copied into contiguous Arrow buffers, and booleans / complex / strings go through the existing VTK conversion and are also copied. Arrow buffers are immutable so consumers cannot mutate VTK memory through the returned table. `to_pandas` consolidates into internal blocks during construction, so the DataFrame is effectively a snapshot. Use `to_arrow` when zero-copy matters. The three behaviors are pinned by tests in `tests/core/test_datasetattributes.py`.
- Error paths: `field_data` raises `ValueError` because its arrays may have differing lengths, name collisions after expansion raise `ValueError`, and an invalid `association` raises `ValueError` via `parse_field_choice`.
- Dependencies: `pandas` and `pyarrow` are optional but added to the test dependencies

I chose not to include the points in the dataframe here and focused on this being for the point/cell data specifically. You can add the points to a point data dataframe via

```py
df = mesh.to_pandas().assign(x=mesh.points[:, 0], y=mesh.points[:, 1], z=mesh.points[:, 2])
```

## Positron caveat

Positron's Data Explorer currently opens objects that pass `isinstance(obj, pd.DataFrame | pl.DataFrame)`. Until Positron adds PyCapsule detection, the workflow is `mesh.point_data.to_pandas()` and then click the result in the Variables pane. Once Positron adopts PyCapsule, `mesh` itself lights up with no further changes.

- Arrow PyCapsule Interface: https://arrow.apache.org/docs/format/CDataInterface/PyCapsuleInterface.html

### CODE REVIEW

**Comment by user27182** [comment]
> Add `to_pandas()`

VTK recently added this API to the new NumPy interface for vtkDataArray: https://gitlab.kitware.com/vtk/vtk/-/merge_requests/13026

how does the new PyVista API compare to this one? Should we mimic this API for ease of maintenance in the future? I.e, can we conditionally use VTK’s `to_pandas()` API directly for VTK 9.7 and later?

**Comment by banesullivan** [comment]
Nice find on [!13026](https://gitlab.kitware.com/vtk/vtk/-/merge_requests/13026), I haven't been paying attention to the work upstream. It changes the follow-up roadmap more than it changes this PR in my opinion

On overlap: VTK's new `FieldDataBase.to_pandas()` uses the exact same `{name}_{i}` expansion we do, so `DataSetAttributes.to_pandas()` lines up semantically. Users see the same behavior whether they call the VTK surface directly or our implementation.

VTK is lacking the Arrow interface, so our `to_arrow()` and `__arrow_c_stream__` are pyvista-only. Zero-copy through the Arrow PyCapsule interface is a different path than going through pandas consolidation, and worth keeping. VTK does add `to_xarray()`, `from_pandas()`, `from_xarray()`, and dict setters on `mesh.point_data` / `cell_data` / `field_data` that we don't have yet so it'll be nice to benefit from those.

Should we delegate on VTK 9.7? IMO, not worth it... This helper isn't much code and it also feeds `to_arrow()` so we keep it regardless, and delegating would make our error contract (the `ValueError` on field data, column collisions, leading-dim mismatches) depend on VTK's behavior at those edges?

Where I think the conditional delegation will pay off is the input side: `from_pandas`, `from_xarray`, and the dict setters, because VTK's input path handles dtype coercion and name normalization that would be annoying to reimplement. I'd queue that up as a follow-up PR referencing [!13026](https://gitlab.kitware.com/vtk/vtk/-/merge_requests/13026) and #5516.

**Comment by user27182** [comment]
> On overlap: VTK's new `FieldDataBase.to_pandas()` uses the exact same `{name}_{i}` expansion we do, so `DataSetAttributes.to_pandas()` lines up semantically. Users see the same behavior whether they call the VTK surface directly or our implementation 

> Should we delegate on VTK 9.7? IMO, not worth it... This helper isn't much code and it also feeds `to_arrow()` so we keep it regardless, and delegating would make our error contract (the `ValueError` on field data, column collisions, leading-dim mismatches) depend on VTK's behavior at those edges?

okay! Sounds reasonable to keep it strictly PyVista code. Maybe we can add a simple test (with `pytest.mark.needs_vtk_version(9, 6, 99)  # >= (9, 7, 0)` to assert identical behavior to VTK’s API for the main use case (no need to test raises or edge cases)?

can use the `vtk-dev-testing` label temporarily for testing, though this test will likely fail for unrelated reasons so we can remove it after confirming the new tests pass.

**Comment by banesullivan** [comment]
Tests related to these changes are passing in the VTK dev job (unrelated failures)

**Comment by user27182** [comment]
> Tests related to these changes are passing in the VTK dev job (unrelated failures)

There are two causes of VTK dev failures, both reported upstream:

https://gitlab.kitware.com/vtk/vtk/-/work_items/20034
https://gitlab.kitware.com/vtk/vtk/-/work_items/20018

I expect both will be resolved soon via new dev wheels, and there is no need to make changes on our end. But I see you already added a patch for the RectilinearGrid 0-points issue. I suggest either patching _both_ issues so that CI for the VTK dev job is green, (you can skip check gc for the vtkBuffer issue, similar to https://github.com/pyvista/pyvista/blob/4c0931617f82d1137f432931b6f3b8701acd07a4/tests/plotting/test_plotting.py#L3637),
or revert the RectilinearGrid 0-points patch since this will be fixed upstream anyway. And then remove the vtk-dev-testing label to skip this job entirely for now.

File: pyvista/core/grid.py [hunk 1]
```diff
@@ -353,6 +353,11 @@ def points(self: Self) -> NumpyArray[float]:
 
         """
         if pv.vtk_version_info >= (9, 4, 0):
+            if self.GetNumberOfPoints() == 0:
+                # VTK >= 9.7 vtkStructuredPointArray returns 3 elements from
+                # __array__ for an empty grid, so vtk_to_numpy fails to reshape
+                # them into (0, 3). Short-circuit to a real empty array.
```
Comment (user27182) [inline pyvista/core/grid.py]:
```suggestion
                # them into (0, 3). Short-circuit to a real empty array.
                # See https://gitlab.kitware.com/vtk/vtk/-/work_items/20034#note_1807104
```
I think we can revert this patch and remove the vtk-dev-testing label to skip this job entirely, or if you want to keep the patch then we should include a link to the upstream issue

**Comment by banesullivan** [comment]
Went ahead and reverted that since it'll likely be fixed upstream. Thanks for tracking those VTK issues!

**Review by akaszynski — APPROVED** [review]
